### PR TITLE
Hopefully fixes unnecessary nightshift announcements

### LINKED ISSUE #36170: Red alerts state that night lighting has been disabled, when night lighting isn't active.
It's only supposed to appear when night lighting is active, but either night lighting is bugged and doesn't show up while active, or somethings wrong with the announcement. Possibly the first one, since I don't see how the relevant bit of code could be wrong:

```
/datum/controller/subsystem/nightshift/proc/check_nightshift(force_set = FALSE)
	var/emergency = GLOB.security_level >= SEC_LEVEL_RED
	var/nightshift = FALSE
	if (!emergency)
		var/time = station_time()
		nightshift = time < nightshift_end_time || time > nightshift_start_time

	var/announcing = TRUE
	if(high_security_mode && !emergency)
		high_security_mode = FALSE
		announce("Restoring night lighting configuration to normal operation.")
		announcing = FALSE
	else if(!high_security_mode && emergency)
		high_security_mode = TRUE
		announce("Night lighting disabled: Station is in a state of emergency.")
		announcing = FALSE

	if((nightshift_active != nightshift) || force_set)
		update_nightshift(nightshift, announcing)

### PR DESCRIPTION
**Pull request by Buggy123**
This _shouldn't_ be necessary but it also _shouldn't_ be announcing that nightshift is disabled when nightshift is off, but it is. Best guess is that somehow vars are carrying over between procs? Is that even a thing?

fixes #36170

### CODE REVIEW

**Comment by Cyberboss** [comment]
@kevinz000

**Comment by ExcessiveUseOfCobblestone** [comment]
can you explain what the issue is

**Comment by Buggy123** [comment]
Right, my bad: #36170

**Comment by ExcessiveUseOfCobblestone** [comment]
You want a single if(nightshift) check below announcing im pretty sure.

**Comment by ChuckTheSheep** [comment]
It should be nightshift_active var, if emergency is TRUE then nightshift is always FALSE. (Why is the logic arranged this way?)

**Comment by Buggy123** [comment]
No idea, I didn't write it.

**Comment by ExcessiveUseOfCobblestone** [comment]
Listen i'm still wondering why 

> announces text
> announce = FALSE

**Comment by ChuckTheSheep** [comment]
@ExcessiveUseOfCobblestone that's to stop the 'update_nightshift()' later on.

Also I was mistaken... the first if() should check for nighttime while the second nighttime_active.

**Comment by ExcessiveUseOfCobblestone** [comment]
This proc is the fountain of ~~youth~~bone hurting juice

**Comment by Buggy123** [comment]
oof

File: code/controllers/subsystem/nightshift.dm [hunk 1]
```diff
@@ -31,14 +31,15 @@ SUBSYSTEM_DEF(nightshift)
...
 	var/announcing = TRUE
-	if(high_security_mode && !emergency)
-		high_security_mode = FALSE
-		announce("Restoring night lighting configuration to normal operation.")
-		announcing = FALSE
-	else if(!high_security_mode && emergency)
-		high_security_mode = TRUE
-		announce("Night lighting disabled: Station is in a state of emergency.")
-		announcing = FALSE
+	if(nightshift_active)
```
Comment (kevinz000) [inline code/controllers/subsystem/nightshift.dm]: Highsec mode should be set regelardless.

**Comment by deathride58** [comment]
You just need to add the nightshift_active check right above the announce procs themselves, not the entire thing that disables nightshift announcements and makes nightshifts unable to trigger.

This "fix" will make it plausible for nightshifts to trigger in the middle of red alert

**Comment by ChuckTheSheep** [comment]
Except it also needs to turn it back on also?

**Comment by Buggy123** [comment]
My bones hurt

**Comment by ChuckTheSheep** [comment]
It's not that bad lol.

Just remember nightshift is the answer to "should it be nightshift?" and nightshift_active is the answer to "Is it nightshift right now?" A check before the two if()s for either one isn't going to work, apply the right one to each if().

**Comment by Buggy123** [comment]
I was more talking about everyone telling me to do different things.

**Comment by ChuckTheSheep** [comment]
You're testing this right? lol

**Comment by Buggy123** [comment]
uuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh.....................................................................

**Comment by ExcessiveUseOfCobblestone** [comment]
`Buggy123:patch-88` you tell me

**Comment by Buggy123** [comment]
HMMMMMMMMMMMM

**Comment by ExcessiveUseOfCobblestone** [comment]
how do you even play the game?

**Comment by Buggy123** [comment]
>My schedule is essentially this:
Mon-thurs
College(+commute): 6:15 A.M. - 5 P.M. (75% of this is free time but with only a chromebook)
All the other hours: sleep and/or playing video games while trying not to think about homework and/or existential dread and trying not to think about how the fundament of life is pointless drudgery
fri-sun
(12 A.M. - 5 A.M.) - (12 P.M. - 3 P.M.): sleep
all the other hours:playing video games while trying not to think about the upcoming work I have to do, and/or existential dread, drudgery, icing, etc

TL;DR i have a very large window where I can use the webeditor to code during the weekday but not test, a very small window where I can use a full PC but I don't want to because I'm tired and sad, and then during the weekend I have plenty of time to do it but I don't want to very much because I'm tired and sad.

**Comment by ChuckTheSheep** [comment]
Hmmm, I don't think it's possible but under 'else if(!high_security_mode && emergency && nightshift)' you should probably ALSO check if '!nightshiftactive' to make sure it's off when it shouldn't be, otherwise it will announce its turning on when its already on. Again shouldn't be possible but you never know.

**Comment by ChuckTheSheep** [comment]
Is it possible for red alert to be turned on before night shift starts, and what would occur?

**Comment by Buggy123** [comment]
Noted, i'll add it.

File: code/controllers/subsystem/nightshift.dm [hunk 2]
```diff
@@ -31,11 +31,11 @@ SUBSYSTEM_DEF(nightshift)
 		nightshift = time < nightshift_end_time || time > nightshift_start_time
 
 	var/announcing = TRUE
-	if(high_security_mode && !emergency)
+	if(high_security_mode && !emergency && nightshift && !nightshift_active)
 		high_security_mode = FALSE
 		announce("Restoring night lighting configuration to normal operation.")
```
Comment (Cyberboss) [inline code/controllers/subsystem/nightshift.dm]: So this message will run if the night shift ISN'T active??

**Comment by kevinz000** [comment]
ill be doing this myself if this isn't fixed /properly/ in about a day because atleast i somewhat test my code and care about my own code.

File: code/controllers/subsystem/nightshift.dm (hunk 2 above)
Comment (Buggy123) [reply code/controllers/subsystem/nightshift.dm]: >Also probably not possible but: if(high_security_mode && !emergency && nightshift) should be if(high_security_mode && !emergency && nightshift && !nightshift_active) to prevent it from making the announcement but not actually doing an update. Again, should not be possible.

goddamn I just do what people tell me to do, yell at them not me

**Comment by Buggy123** [comment]
>ill be doing this myself if this isn't fixed /properly/ in about a day because atleast i somewhat test my code and care about my own code.

What's wrong with it?

**Comment by ExcessiveUseOfCobblestone** [comment]
You having to ask what's wrong with it.

**Comment by Buggy123** [comment]
But... it works, there isn't a obscene number of commits (though if five is still too much I can squash it), and it isn't shitcode. I see nothing wrong with it.

File: code/controllers/subsystem/nightshift.dm (hunk 2 above)
Comment (ChuckTheSheep) [reply code/controllers/subsystem/nightshift.dm]: If high security mode is on AND it is not an emergency AND it SHOULD be nightshift AND it is currently NOT, THEN make the announcement cause if by some miracle nightshift_active is already on later on in the code it won't do anything in the update but that announcement will still go off.

Honestly announcement should be a text argument that gets modified.

Comment (ExcessiveUseOfCobblestone) [reply code/controllers/subsystem/nightshift.dm]: This proc needs comments or the vars need to be better labeled, this is a headache for seemingly everyone.

Comment (ChuckTheSheep) [reply code/controllers/subsystem/nightshift.dm]: Yeah just realized, it does goof up by doing this... ugh.

**Comment by ChuckTheSheep** [comment]
Nah it's goof'd. Not you fault though honestly.

**Comment by ExcessiveUseOfCobblestone** [comment]
is high_security_mode just nightshift "on" with normal lighting?

**Comment by ChuckTheSheep** [comment]
High sec mode is a var under night shift that updates when there is an emergency, the issue is all the extra && in the ifs.

**Comment by ChuckTheSheep** [comment]
Also, am I mistaken or are the actual default "good morning/evening" announcements reversed?

**Comment by ChuckTheSheep** [comment]
Tested and works.

Only one thing I'm still trying to get is a message for when it is already red alert but night end comes, so maybe say something like 'night shift lighting is delayed' but its not that big of a deal.

Feel free to remove extra line breaks.

`/datum/controller/subsystem/nightshift/proc/check_nightshift(force_set = FALSE)
	var/emergency = GLOB.security_level >= SEC_LEVEL_RED
	var/announcing = TRUE
	var/time = station_time()
	var/night_time = ((time < nightshift_end_time) || (time > nightshift_start_time))
	
	if(high_security_mode != emergency)
		high_security_mode = emergency
		if(night_time)
			if(!emergency)
				announcing = "Restoring night lighting configuration to normal operation."
			else
				announcing = "Disabling night lighting: Station is in a state of emergency."
	
	if(emergency)
		night_time = FALSE

	if((nightshift_active != night_time) || force_set)
		update_nightshift(night_time, announcing)	
	
/datum/controller/subsystem/nightshift/proc/update_nightshift(active, announce = TRUE)
	nightshift_active = active
	if(announce)
		if(istext(announce))
			announce(announce)
		else
			if (active)
				announce("Good evening, crew. To reduce power consumption and stimulate the circadian rhythms of some species, all of the lights aboard the station have been dimmed for the night.")
			else
				announce("Good morning, crew. As it is now day time, all of the lights aboard the station have been restored to their former brightness.")
	for(var/A in GLOB.apcs_list)
		var/obj/machinery/power/apc/APC = A
		if (APC.area && (APC.area.type in GLOB.the_station_areas))
			APC.set_nightshift(active)
			CHECK_TICK`

**Comment by Buggy123** [comment]
Thank you chuck, that's _much_ better.

**Review by duncathan — CHANGES_REQUESTED** [review]
use tabs not spaces

**Comment by Buggy123** [comment]
There are spaces there? Fuck.
